pkg/network/encoding/kafka.go

```diff
@@ -6,140 +6,113 @@
+			// right because we're re-using it multiple times and should quickly
+			// converge to a final size after a few calls to
+			// `GetKafkaAggregations`
+			KafkaAggregations: make([]*model.KafkaAggregation, 0, 10),
+		},
+		byConnection: GroupByConnection("kafka", payload.Kafka, func(key kafka.Key) types.ConnectionKey {
+			return key.ConnectionKey
+		}),
 	}
-
-	if c.SPort == a.dport && c.DPort == a.sport {
-		// We have a collision with another `ConnectionStats`, but this is a
-		// legit scenario where we're dealing with the opposite ends of the
-		// same connection, which means both server and client are in the same host.
-		// In this particular case it is correct to have both connections
-		// (client:server and server:client) referencing the same Kafka data.
-		return a.DataStreamsAggregations
-	}
-
-	// Return nil otherwise. This is to prevent multiple `ConnectionStats` with
-	// exactly the same source and destination addresses but different PIDs to
-	// "bind" to the same DataStreamsAggregations object, which would result in a
-	// overcount problem. (Note that this is due to the fact that
-	// `types.ConnectionKey` doesn't have a PID field.) This happens mostly in the
-	// context of pre-fork web servers, where multiple worker proceses share the
-	// same socket
-	return nil
 }
 
-func newKafkaEncoder(payload *network.Connections) *kafkaEncoder {
-	if len(payload.Kafka) == 0 {
+func (e *kafkaEncoder) GetKafkaAggregations(c network.ConnectionStats) []byte {
+	if e == nil {
 		return nil
 	}
 
-	encoder := &kafkaEncoder{
-		aggregations: make(map[types.ConnectionKey]*kafkaAggregationWrapper, len(payload.Conns)),
+	connectionData := e.byConnection.Find(c)
+	if connectionData == nil || len(connectionData.Data) == 0 || connectionData.IsPIDCollision(c) {
+		return nil
 	}
 
-	// pre-populate aggregation map with keys for all existent connections
-	// this allows us to skip encoding orphan Kafka objects that can't be matched to a connection
-	for _, conn := range payload.Conns {
-		for _, key := range network.ConnectionKeysFromConnectionStats(conn) {
-			if log.ShouldLog(seelog.TraceLvl) {
-				log.Tracef("Payload has a connection %v and was converted to kafka key %v", conn, key)
-			}
-			encoder.aggregations[key] = nil
-		}
-	}
-	encoder.buildAggregations(payload)
-	return encoder
+	return e.encodeData(connectionData)
 }
 
-func (e *kafkaEncoder) GetKafkaAggregations(c network.ConnectionStats) *model.DataStreamsAggregations {
+func (e *kafkaEncoder) Close() {
 	if e == nil {
-		return nil
+		return
 	}
 
-	connectionKeys := network.ConnectionKeysFromConnectionStats(c)
-	for _, key := range connectionKeys {
-		if aggregation := e.aggregations[key]; aggregation != nil {
-			return e.aggregations[key].ValueFor(c)
-		}
+	e.reset()
+	e.byConnection.Close()
+}
+
+func (e *kafkaEncoder) encodeData(connectionData *USMConnectionData[kafka.Key, *kafka.RequestStat]) []byte {
+	e.reset()
```

Comment:
Why do we need to call reset if we're initialising the `kafkaAggregation` inside the for loop?